src/dep.cc

```diff
@@ -249,13 +252,26 @@ struct RuleMerger {
         n->loc = r->loc;
     }
 
+    std::set<string> outs;
+    SymbolSet all_outputs = SymbolSet();
+    all_outputs.insert(output);
+
     for (auto& implicit_output : implicit_outputs) {
       n->implicit_outputs.push_back(implicit_output.first);
+      all_outputs.insert(implicit_output.first);
       for (const Rule* r : implicit_output.second->rules) {
         FillDepNodeFromRule(output, r, n);
       }
     }
 
+    for (auto& symlink_output : symlink_outputs) {
+      if (!all_outputs.exists(symlink_output)) {
+        ERROR_LOC(primary_rule->cmd_loc(), "*** undeclared symlink output: %s",
+                  symlink_output.c_str());
+      }
+      n->symlink_outputs.push_back(symlink_output);
```

Comment:
nit: maybe remove `symlink_output` from `all_outputs` so that repeating an item in the list of symlink outputs is an error

Reply:
SymbolSet doesn't have a delete API, so I'll skip doing this for now to keep this PR scoped.